fetch/fetch.go

```diff
@@ -178,68 +219,121 @@ func (f *Fetcher) FetchChainData(ctx context.Context) error {
-						// as opposed to this error being a show-stopper for the fetcher
-						f.logger.Error("unable to save block", zap.String("err", saveErr.Error()))
+		f.logger.Debug("Added block data to batch", zap.Int64("number", block.Height))
 
-						continue
-					}
+		// Get block results
+		txResults := results[blockIndex]
 
-					f.logger.Debug("Added block data to batch", zap.Int64("number", block.Height))
+		// Save the fetched transaction results
+		for _, txResult := range txResults {
+			if err := wb.SetTx(txResult); err != nil {
+				f.logger.Error("unable to  save tx", zap.String("err", err.Error()))
 
-					// Get block results
-					txResults := item.chunk.results[blockIndex]
+				continue
+			}
 
-					// Save the fetched transaction results
-					for _, txResult := range txResults {
-						if err := wb.SetTx(txResult); err != nil {
-							f.logger.Error("unable to  save tx", zap.String("err", err.Error()))
+			f.logger.Debug(
+				"Added tx to batch",
+				zap.String("hash", base64.StdEncoding.EncodeToString(txResult.Tx.Hash())),
+			)
+		}
 
-							continue
-						}
+		// Alert any listeners of a new saved block
+		event := &types.NewBlock{
+			Block:   block,
+			Results: txResults,
+		}
 
-						f.logger.Debug(
-							"Added tx to batch",
-							zap.String("hash", base64.StdEncoding.EncodeToString(txResult.Tx.Hash())),
-						)
-					}
+		f.events.SignalEvent(event)
+	}
 
-					// Alert any listeners of a new saved block
-					event := &types.NewBlock{
-						Block:   block,
-						Results: txResults,
-					}
+	f.logger.Info(
+		"Added to batch block and tx data for range",
+		zap.Uint64("from", from),
+		zap.Uint64("to", to),
+	)
 
-					f.events.SignalEvent(event)
-				}
+	// Save the latest height data
+	if err := wb.SetLatestHeight(to); err != nil {
+		if rErr := wb.Rollback(); rErr != nil {
+			return fmt.Errorf("unable to save latest height info, %w, %w", err, rErr)
+		}
 
-				f.logger.Info(
-					"Added to batch block and tx data for range",
-					zap.Uint64("from", item.chunkRange.from),
-					zap.Uint64("to", item.chunkRange.to),
-				)
+		return fmt.Errorf("unable to save latest height info, %w", err)
+	}
 
-				// Save the latest height data
-				if err := wb.SetLatestHeight(item.chunkRange.to); err != nil {
-					if rErr := wb.Rollback(); rErr != nil {
-						return fmt.Errorf("unable to save latest height info, %w, %w", err, rErr)
-					}
+	if err := wb.Commit(); err != nil {
+		return fmt.Errorf("error persisting block information into storage, %w", err)
+	}
 
-					return fmt.Errorf("unable to save latest height info, %w", err)
-				}
+	return nil
+}
 
-				if err := wb.Commit(); err != nil {
-					return fmt.Errorf("error persisting block information into storage, %w", err)
-				}
-			}
+func getGenesisBlock(client Client) (*bft_types.Block, error) {
+	gblock, err := client.GetGenesis()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get genesis block: %w", err)
+	}
+
+	if gblock.Genesis == nil {
```

Comment:
the `Client` is an interface, there could be a bad implementation that returns a malformed type, I think it's better to fail gracefully instead of crashing with nil deref